File: java/Samples/Biometrics/VB/SimpleFacesSampleVB/EnrollFromCamera.vb
Imports System
Imports System.IO
Imports System.Windows.Forms
Imports Neurotec.Biometrics
Imports Neurotec.Biometrics.Client
Imports Neurotec.Devices
Imports Neurotec.Images

Partial Public Class EnrollFromCamera
	Inherits UserControl
#Region "Public constructor"

	Public Sub New()
		InitializeComponent()
	End Sub

#End Region

#Region "Private fields"

	Private _biometricClient As NBiometricClient
	Private _subject As NSubject
	Private _deviceManager As NDeviceManager

#End Region

#Region "Public properties"

	Public Property BiometricClient() As NBiometricClient
		Get
			Return _biometricClient
		End Get
		Set(ByVal value As NBiometricClient)
			_biometricClient = value
		End Set
	End Property

#End Region

#Region "Private methods"

	Private Sub UpdateCameraList()
		cbCameras.BeginUpdate()
		Try
			cbCameras.Items.Clear()
			For Each device As NDevice In _deviceManager.Devices
				cbCameras.Items.Add(device)
			Next device

			If _biometricClient.FaceCaptureDevice Is Nothing AndAlso cbCameras.Items.Count > 0 Then
				cbCameras.SelectedIndex = 0
				Return
			End If

			If _biometricClient.FaceCaptureDevice IsNot Nothing Then
				cbCameras.SelectedIndex = cbCameras.Items.IndexOf(_biometricClient.FaceCaptureDevice)
			End If
		Finally
			cbCameras.EndUpdate()
		End Try
	End Sub

	Private Sub EnableControls(ByVal capturing As Boolean)
		Dim hasTemplate = (Not capturing) AndAlso _subject IsNot Nothing AndAlso _subject.Status = NBiometricStatus.Ok
		btnSaveImage.Enabled = hasTemplate
		btnSaveTemplate.Enabled = hasTemplate
		btnStart.Enabled = Not capturing
		btnRefreshList.Enabled = Not capturing
		btnStop.Enabled = capturing
		cbCameras.Enabled = Not capturing
		btnStartExtraction.Enabled = capturing AndAlso Not chbCaptureAutomatically.Checked
		chbCaptureAutomatically.Enabled = Not capturing
		chbCheckLiveness.Enabled = Not capturing
	End Sub

	Private Sub OnCapturingCompleted(ByVal r As IAsyncResult)
		If InvokeRequired Then
			BeginInvoke(New AsyncCallback(AddressOf OnCapturingCompleted), r)
		Else
			Try
				Dim status As NBiometricStatus = _biometricClient.EndCapture(r)
				' If Stop button was pushed
				If status = NBiometricStatus.Canceled Then
					Return
				End If

				lblStatus.Text = status.ToString()
				If status <> NBiometricStatus.Ok Then
					' Since capture failed start capturing again
					_subject.Faces(0).Image = Nothing
					_biometricClient.BeginCapture(_subject, AddressOf OnCapturingCompleted, Nothing)
				Else
					EnableControls(False)
				End If
			Catch ex As Exception
				Utils.ShowException(ex)
				lblStatus.Text = String.Empty
				EnableControls(False)
			End Try
		End If
	End Sub

#End Region

#Region "Private form events"

	Private Sub EnrollFromCameraLoad(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
		If (Not DesignMode) Then
			Try
				lblStatus.Text = String.Empty
				_deviceManager = _biometricClient.DeviceManager
				saveImageDialog.Filter = NImages.GetSaveFileFilterString()
				UpdateCameraList()
			Catch ex As Exception
				Utils.ShowException(ex)
			End Try
		End If
	End Sub

	Private Sub BtnRefreshListClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnRefreshList.Click
		UpdateCameraList()
	End Sub

	Private Sub CbCameraSelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles cbCameras.SelectedIndexChanged
		_biometricClient.FaceCaptureDevice = TryCast(cbCameras.SelectedItem, NCamera)
	End Sub

	Private Sub BtnStartClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnStart.Click
		If _biometricClient.FaceCaptureDevice Is Nothing Then
			MessageBox.Show("Please select camera from the list")
			Return
		End If
		' Set face capture from stream
		Dim face As New NFace() With {.CaptureOptions = NBiometricCaptureOptions.Stream}
		If Not chbCaptureAutomatically.Checked Then
			face.CaptureOptions = face.CaptureOptions Or NBiometricCaptureOptions.Manual
		End If
		_subject = New NSubject()
		_subject.Faces.Add(face)
		facesView.Face = face

		' Begin capturing faces
		_biometricClient.BeginCapture(_subject, AddressOf OnCapturingCompleted, Nothing)

		lblStatus.Text = String.Empty
		EnableControls(True)
	End Sub

	Private Sub BtnStopClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnStop.Click
		_biometricClient.Cancel()
		EnableControls(False)
	End Sub

	Private Sub BtnSaveTemplateClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnSaveTemplate.Click
		If saveTemplateDialog.ShowDialog() = DialogResult.OK Then
			File.WriteAllBytes(saveTemplateDialog.FileName, _subject.GetTemplateBuffer().ToArray())
		End If
	End Sub

	Private Sub BtnStartExtractionClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnStartExtraction.Click
		lblStatus.Text = "Extracting ..."
		' Begin extraction
		_biometricClient.ForceStart()
	End Sub

	Private Sub BtnSaveImageClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnSaveImage.Click
		If saveImageDialog.ShowDialog() = DialogResult.OK Then
			_subject.Faces(0).Image.Save(saveImageDialog.FileName)
		End If
	End Sub

	Private Sub EnrollFromCameraVisibleChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.VisibleChanged
		If _biometricClient IsNot Nothing Then
			EnableControls(False)
		End If
	End Sub

	Private Sub ChbCheckLivenessCheckedChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles chbCheckLiveness.CheckedChanged
		_biometricClient.FacesLivenessMode = If(chbCheckLiveness.Checked, NLivenessMode.PassiveAndActive, NLivenessMode.None)
	End Sub

#End Region

End Class

File: java/Samples/Biometrics/CS/SimpleFacesSampleCS/EnrollFromCamera.cs
using System;
using System.IO;
using System.Windows.Forms;
using Neurotec.Biometrics;
using Neurotec.Biometrics.Client;
using Neurotec.Devices;
using Neurotec.Images;

namespace Neurotec.Samples
{
	public partial class EnrollFromCamera : UserControl
	{
		#region Public constructor

		public EnrollFromCamera()
		{
			InitializeComponent();
		}

		#endregion

		#region Private fields

		private NBiometricClient _biometricClient;
		private NSubject _subject;
		private NDeviceManager _deviceManager;

		#endregion

		#region Public properties

		public NBiometricClient BiometricClient
		{
			get { return _biometricClient; }
			set { _biometricClient = value; }
		}

		#endregion

		#region Private methods

		private void UpdateCameraList()
		{
			cbCameras.BeginUpdate();
			try
			{
				cbCameras.Items.Clear();
				foreach (NDevice device in _deviceManager.Devices)
				{
					cbCameras.Items.Add(device);
				}

				if (_biometricClient.FaceCaptureDevice == null && cbCameras.Items.Count > 0)
				{
					cbCameras.SelectedIndex = 0;
					return;
				}

				if (_biometricClient.FaceCaptureDevice != null)
				{
					cbCameras.SelectedIndex = cbCameras.Items.IndexOf(_biometricClient.FaceCaptureDevice);
				}
			}
			finally
			{
				cbCameras.EndUpdate();
			}
		}

		private void EnableControls(bool capturing)
		{
			var hasTemplate = !capturing && _subject != null && _subject.Status == NBiometricStatus.Ok;
			btnSaveImage.Enabled = hasTemplate;
			btnSaveTemplate.Enabled = hasTemplate;
			btnStart.Enabled = !capturing;
			btnRefreshList.Enabled = !capturing;
			btnStop.Enabled = capturing;
			cbCameras.Enabled = !capturing;
			btnStartExtraction.Enabled = capturing && !chbCaptureAutomatically.Checked;
			chbCaptureAutomatically.Enabled = !capturing;
			chbCheckLiveness.Enabled = !capturing;
		}

		private void OnCapturingCompleted(IAsyncResult r)
		{
			if (InvokeRequired)
			{
				BeginInvoke(new AsyncCallback(OnCapturingCompleted), r);
			}
			else
			{
				try
				{
					NBiometricStatus status = _biometricClient.EndCapture(r);
					// If Stop button was pushed
					if (status == NBiometricStatus.Canceled) return;

					lblStatus.Text = status.ToString();
					if (status != NBiometricStatus.Ok)
					{
						// Since capture failed start capturing again
						_subject.Faces[0].Image = null;
						_biometricClient.BeginCapture(_subject, OnCapturingCompleted, null);
					}
					else
					{
						EnableControls(false);
					}
				}
				catch (Exception ex)
				{
					Utils.ShowException(ex);
					lblStatus.Text = string.Empty;
					EnableControls(false);
				}
			}
		}

		#endregion

		#region Private form events

		private void EnrollFromCameraLoad(object sender, EventArgs e)
		{
			if (!DesignMode)
			{
				try
				{
					lblStatus.Text = string.Empty;
					_deviceManager = _biometricClient.DeviceManager;
					saveImageDialog.Filter = NImages.GetSaveFileFilterString();
					UpdateCameraList();
				}
				catch (Exception ex)
				{
					Utils.ShowException(ex);
				}
			}
		}

		private void BtnRefreshListClick(object sender, EventArgs e)
		{
			UpdateCameraList();
		}

		private void CbCameraSelectedIndexChanged(object sender, EventArgs e)
		{
			_biometricClient.FaceCaptureDevice = cbCameras.SelectedItem as NCamera;
		}

		private void BtnStartClick(object sender, EventArgs e)
		{
			if (_biometricClient.FaceCaptureDevice == null)
			{
				MessageBox.Show(@"Please select camera from the list");
				return;
			}
			// Set face capture from stream
			var face = new NFace { CaptureOptions = NBiometricCaptureOptions.Stream };
			if (!chbCaptureAutomatically.Checked) face.CaptureOptions |= NBiometricCaptureOptions.Manual;
			_subject = new NSubject();
			_subject.Faces.Add(face);
			facesView.Face = face;

			// Begin capturing faces
			_biometricClient.BeginCapture(_subject, OnCapturingCompleted, null);

			lblStatus.Text = string.Empty;
			EnableControls(true);
		}

		private void BtnStopClick(object sender, EventArgs e)
		{
			_biometricClient.Cancel();
			EnableControls(false);
		}

		private void BtnSaveTemplateClick(object sender, EventArgs e)
		{
			if (saveTemplateDialog.ShowDialog() == DialogResult.OK)
			{
				File.WriteAllBytes(saveTemplateDialog.FileName, _subject.GetTemplateBuffer().ToArray());
			}
		}

		private void BtnStartExtractionClick(object sender, EventArgs e)
		{
			lblStatus.Text = @"Extracting ...";
			// Begin extraction
			_biometricClient.ForceStart();
		}

		private void BtnSaveImageClick(object sender, EventArgs e)
		{
			if (saveImageDialog.ShowDialog() == DialogResult.OK)
			{
				_subject.Faces[0].Image.Save(saveImageDialog.FileName);
			}
		}

		private void EnrollFromCameraVisibleChanged(object sender, EventArgs e)
		{
			if (Visible && _biometricClient != null)
			{
				EnableControls(false);
			}
		}

		private void ChbCheckLivenessCheckedChanged(object sender, EventArgs e)
		{
			// set liveness check mode
			_biometricClient.FacesLivenessMode = chbCheckLiveness.Checked ? NLivenessMode.PassiveAndActive : NLivenessMode.None;
		}

		#endregion

	}
}
